src/theory/tools/wp/bir_htScript.sml

```diff
@@ -0,0 +1,614 @@
+
+open bir_wpTheory;
+
+open HolBACoreSimps;
+
+val _ = new_theory "bir_ht";
+
+val bir_stmtb_is_not_assume_def = Define `
+  (bir_stmtb_is_not_assume (BStmt_Declare v) = T) /\
+  (bir_stmtb_is_not_assume (BStmt_Assign v ex) = T) /\
+  (bir_stmtb_is_not_assume (BStmt_Assert ex) = T) /\
+  (bir_stmtb_is_not_assume (BStmt_Assume ex) = F) /\
+  (bir_stmtb_is_not_assume (BStmt_Observe ec el obf) = T)
+`;
+
+val bir_stmtb_not_assume_never_assviol =
+  store_thm("bir_stmtb_not_assume_never_assviol",
+  ``!stmtb st obs st'.
+    bir_stmtb_is_not_assume stmtb ==>
+    (st.bst_status <> BST_AssumptionViolated) ==>
+    (bir_exec_stmtB stmtb st = (obs, st')) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+REPEAT STRIP_TAC >>
+Cases_on `stmtb` >> (
+  FULL_SIMP_TAC std_ss [bir_stmtb_is_not_assume_def,
+                        bir_exec_stmtB_def]
+) >| [
+  FULL_SIMP_TAC std_ss [bir_exec_stmt_declare_def] >>
+  Cases_on `bir_env_varname_is_bound st.bst_environ
+              (bir_var_name b)` >| [
+    FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+    FULL_SIMP_TAC std_ss [] >>
+    Cases_on `bir_env_update (bir_var_name b)
+                (bir_declare_initial_value (bir_var_type b))
+                (bir_var_type b)
+                st.bst_environ` >| [
+      FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+      RW_TAC std_ss [] >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+      FULL_SIMP_TAC std_ss [] >>
+      RW_TAC std_ss [] >>
+      FULL_SIMP_TAC (std_ss++holBACore_ss) []
+    ]
+  ],
+
+  FULL_SIMP_TAC std_ss [bir_exec_stmt_assign_def, LET_DEF] >>
+  Cases_on `bir_env_write b (bir_eval_exp b0 st.bst_environ)
+                          st.bst_environ` >| [
+    FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) []
+  ],
+
+  FULL_SIMP_TAC std_ss [bir_exec_stmt_assert_def] >>
+  Cases_on `bir_dest_bool_val (bir_eval_exp b st.bst_environ)` >| [
+    FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+    Cases_on `x` >> (
+      FULL_SIMP_TAC (std_ss++holBACore_ss) []
+    ) >>
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) []
+  ],
+
+  FULL_SIMP_TAC std_ss [bir_exec_stmt_observe_def] >>
+  Cases_on `bir_dest_bool_val (bir_eval_exp b st.bst_environ)` >| [
+    FULL_SIMP_TAC std_ss [bir_state_set_failed_def] >>
+    RW_TAC std_ss [] >>
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+    Cases_on `x` >> (
+      FULL_SIMP_TAC (std_ss++holBACore_ss) []
+    )
+  ]
+]
+);
+
+val bir_stmtsB_has_no_assumes_def = Define `
+  (bir_stmtsB_has_no_assumes [] = T) /\
+  (bir_stmtsB_has_no_assumes (h::t) =
+    (bir_stmtb_is_not_assume h) /\ (bir_stmtsB_has_no_assumes t)
+  )
+`;
+
+(* TODO: Move this elsewhere... *)
```

Comment:
Can you explain why in the TODO? Same for the following ones.

Reply:
The reason for this particular TODO is that the lemma is very general and as such does not belong in such a specialized theory as `bir_htTheory`. This theorem deals with fundamental properties of `bir_exec_stmtB` and IMO should be put in the same theory where this is defined (`bir_programTheory`). Similar for the following, but different theories.

The general thought has been detailed in a reply to your main comment.

Reply:
I saw the reply to the main comment, no need to reexplain it here. This comment will be remarked as outdated if you do the spreading-out commit.